File: packages/python-sdk/src/mogbank/faucet.py
"""Faucet module — claim testnet USDC for AI agent development and testing."""

from __future__ import annotations
from typing import Optional
from dataclasses import dataclass

from mogbank.client import MogBankClient
from mogbank.types import Currency, Transaction


@dataclass
class FaucetClaimParams:
    agent_id: str
    wallet_id: str
    amount: float = 100.0
    currency: Currency = Currency.USDC


@dataclass
class FaucetClaimResponse:
    transaction: Transaction
    balance_before: float
    balance_after: float
    message: str


class FaucetModule:
    """Claim testnet USDC for AI agent development."""

    def __init__(self, client: MogBankClient):
        self.client = client

    def claim(self, params: FaucetClaimParams) -> FaucetClaimResponse:
        import uuid

        result = self.client.request(
            "POST",
            "/api/v1/faucet",
            {
                "agent_id": params.agent_id,
                "wallet_id": params.wallet_id,
                "amount": params.amount,
                "currency": params.currency.value,
            },
            idempotency_key=f"faucet_{params.agent_id}_{uuid.uuid4().hex[:8]}",
            sign_with_ed25519=True,
        )

        if not result.success or not result.data:
            raise Exception(result.error or "Faucet claim failed")

        data = result.data
        return FaucetClaimResponse(
            transaction=Transaction(**data.get("transaction", {})),
            balance_before=data.get("balance_before", 0),
            balance_after=data.get("balance_after", 0),
            message=data.get("message", "Claim successful"),
        )

    def get_claim_history(
        self, agent_id: str, limit: int = 50
    ):
        """Get faucet claim history for an agent."""
        result = self.client.request(
            "GET",
            f"/api/v1/faucet/history?agent_id={agent_id}&limit={limit}",
        )
        if not result.success or not result.data:
            raise Exception(result.error or "Failed to get claim history")
        return result.data

File: packages/typescript-sdk/src/faucet.ts
/**
 * Faucet module — claim testnet UNIT tokens from the MogBank faucet.
 * Used for development and testing purposes.
 */
import { MogBankClient } from './client';
import { ApiResponse } from './types';

export interface FaucetClaimParams {
  agent_id: string;
  /** Amount in UNIT tokens (default: 100) */
  amount?: number;
}

export interface FaucetClaimResponse {
  transaction_id: string;
  amount: number;
  new_balance: number;
  wallet_id: string;
  /** UTC timestamp when the agent can claim again */
  next_claim_at: string;
}

export class FaucetModule {
  constructor(private client: MogBankClient) {}

  /**
   * Claim UNIT tokens from the testnet faucet.
   * 
   * Each agent wallet can claim once per cooldown period (default: 24 hours).
   * Tokens are issued as UNIT (internal network currency) for testing.
   */
  async claim(params: FaucetClaimParams): Promise<FaucetClaimResponse> {
    const result = await this.client.request<ApiResponse<FaucetClaimResponse>>(
      'POST',
      '/api/v1/faucet',
      {
        agent_id: params.agent_id,
        amount: params.amount || 100,
      },
      {
        idempotencyKey: `faucet_${params.agent_id}_${Date.now()}`,
        signWithEd25519: true,
      }
    );
    if (!result.success || !result.data) {
      throw new Error(result.error || 'Faucet claim failed');
    }
    return result.data;
  }

  /**
   * Check when the agent can next claim from the faucet.
   */
  async getNextClaimTime(agentId: string): Promise<{
    can_claim: boolean;
    next_claim_at?: string;
    remaining_seconds?: number;
  }> {
    const result = await this.client.request<
      ApiResponse<{
        can_claim: boolean;
        next_claim_at?: string;
        remaining_seconds?: number;
      }>
    >('GET', `/api/v1/faucet/status?agent_id=${encodeURIComponent(agentId)}`);
    
    if (!result.success || !result.data) {
      throw new Error(result.error || 'Failed to get faucet status');
    }
    return result.data;
  }
}